pkg/util/cloudfoundry/bbscache.go

```diff
@@ -193,38 +210,76 @@ func (bc *BBSCache) readData() {
 	defer bc.Unlock()
 	log.Debug("Data from BBS API read successfully, refreshing the cache")
 	bc.actualLRPs = actualLRPs
+	bc.actualLRPByInstanceGUID = actualLRPByInstanceGUID
 	bc.desiredLRPs = desiredLRPs
 	bc.lastUpdated = time.Now()
 	bc.pollSuccesses++
 }
 
-func (bc *BBSCache) readActualLRPs() (map[string][]ActualLRP, error) {
+func (bc *BBSCache) readActualLRPs() (map[string][]ActualLRP, map[string]*ActualLRP, error) {
 	actualLRPs := map[string][]ActualLRP{}
+	actualLRPByInstanceGUID := map[string]*ActualLRP{}
 	actualLRPsBBS, err := bc.bbsAPIClient.ActualLRPs(bc.bbsAPIClientLogger, models.ActualLRPFilter{})
 	if err != nil {
-		return actualLRPs, err
+		return actualLRPs, actualLRPByInstanceGUID, err
 	}
 	for _, lrp := range actualLRPsBBS {
 		alrp := ActualLRPFromBBSModel(lrp)
 		if lrpList, ok := actualLRPs[alrp.AppGUID]; ok {
 			actualLRPs[alrp.AppGUID] = append(lrpList, alrp)
+			actualLRPByInstanceGUID[alrp.InstanceGUID] = &actualLRPs[alrp.AppGUID][len(lrpList)]
 		} else {
 			actualLRPs[alrp.AppGUID] = []ActualLRP{alrp}
+			actualLRPByInstanceGUID[alrp.InstanceGUID] = &actualLRPs[alrp.AppGUID][0]
 		}
 	}
 	log.Debugf("Successfully read %d Actual LRPs", len(actualLRPsBBS))
-	return actualLRPs, nil
+	return actualLRPs, actualLRPByInstanceGUID, nil
 }
 
-func (bc *BBSCache) readDesiredLRPs() ([]DesiredLRP, error) {
+func (bc *BBSCache) readDesiredLRPs() (map[string]DesiredLRP, error) {
 	desiredLRPsBBS, err := bc.bbsAPIClient.DesiredLRPs(bc.bbsAPIClientLogger, models.DesiredLRPFilter{})
 	if err != nil {
-		return []DesiredLRP{}, err
+		return map[string]DesiredLRP{}, err
 	}
-	desiredLRPs := make([]DesiredLRP, len(desiredLRPsBBS))
-	for i, lrp := range desiredLRPsBBS {
-		desiredLRPs[i] = DesiredLRPFromBBSModel(lrp)
+	desiredLRPs := make(map[string]DesiredLRP, len(desiredLRPsBBS))
+	for _, lrp := range desiredLRPsBBS {
+		desiredLRP := DesiredLRPFromBBSModel(lrp)
+		desiredLRPs[desiredLRP.AppGUID] = desiredLRP
 	}
 	log.Debugf("Successfully read %d Desired LRPs", len(desiredLRPsBBS))
 	return desiredLRPs, nil
 }
+
+// ExtractTags extract all the container tags for each app in the cache running on the specified node
+// and returns a mapping of tags by instance GUID
+func (bc *BBSCache) ExtractTags(nodename string) map[string][]string {
+	tags := map[string][]string{}
+	alrps, dlrps := bc.GetAllLRPs()
```

Comment:
does the bbs api offer the possibility to query LRPs by node name directly? might be super useful to avoid extra processing here

Reply:
We're not reading from the BBS API here, but from our local cache. That said, I can probably separate by cell ID when I read all the actual LRPs and store them in our cache for easier lookup

Reply:
sounds good 👍